src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,414 @@
+            }
+
+            static SourceText GetIndentedText(IndentationResult indentation, SourceText text, DocumentOptionSet documentOptions)
+            {
+                var indentText = indentation.Offset.CreateIndentationString(documentOptions.GetOption(FormattingOptions.UseTabs), documentOptions.GetOption(FormattingOptions.TabSize));
+                return text.WithChanges(new TextChange(new TextSpan(indentation.BasePosition, 0), indentText));
+            }
+        }
+
+        public override async Task<bool> IsValidForBraceCompletionAsync(char brace, int openingPosition, Document document, CancellationToken cancellationToken)
+        {
+            // Only potentially valid for curly brace completion if not in an interpolation brace completion context.
+            if (OpeningBrace == brace && await InterpolationBraceCompletionService.IsContextAsync(document, openingPosition, cancellationToken).ConfigureAwait(false))
+            {
+                return false;
+            }
+
+            return await base.IsValidForBraceCompletionAsync(brace, openingPosition, document, cancellationToken).ConfigureAwait(false);
+        }
+
+        private bool ContainsOnlyWhitespace(SourceText text, int openingPosition, int closingPosition)
+        {
+            var start = openingPosition;
+            start = text[start] == OpeningBrace ? start + 1 : start;
+
+            var end = closingPosition - 1;
+            end = text[end] == ClosingBrace ? end - 1 : end;
+
+            if (!PositionInSnapshot(start, text) ||
+                !PositionInSnapshot(end, text))
+            {
+                return false;
+            }
+
+            for (var i = start; i <= end; i++)
+            {
+                if (!char.IsWhiteSpace(text[i]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+
+            static int GetValueInValidRange(int value, int smallest, int largest)
+            => Math.Max(smallest, Math.Min(value, largest));
+
+            static bool PositionInSnapshot(int position, SourceText text)
+                => GetValueInValidRange(position, 0, Math.Max(0, text.Length - 1)) == position;
+        }
+
+        private static async Task<TextChange> GetCollapsedTextChangeAsync(SourceText newDocumentText, Document oldDocument, CancellationToken cancellationToken)
+        {
+            var newDocument = oldDocument.WithText(newDocumentText);
+            var changes = await newDocument.GetTextChangesAsync(oldDocument, cancellationToken).ConfigureAwait(false);
+            return Collapse(newDocumentText, changes.ToImmutableArray());
+        }
+
+        private static async Task<SourceText> FormatTrackingSpanAsync(Document document, int openingPoint, int closingPoint, bool shouldHonorAutoFormattingOnCloseBraceOption, CancellationToken cancellationToken, ImmutableArray<AbstractFormattingRule>? rules = null)
```

Comment:
this is from the original method in CurlyBraceCompletionSession.FormatTrackingSpan, with extra logic added to apply the formatting to the document instead of editor subject buffer at the end.